### PR TITLE
add link tester, issue #151

### PR DESCRIPTION
**Pull request by arve0**
Might want to do 

```
"test:links": "yarn run build && node test-links.js"
```

to assure a build to `dist` before checking links.

Currently 43 broken links:
```
Checking all links found at http://localhost:3000
.........................................................................xx.x..................................................................................t......x..............................................x.............x........................................................x.....x..x......x................x...x......x...x............x.................................................xx...........................x.....x...x...x.......x....x.........................................................x...................................................................................................................................xx...............................................................................................x.xx......xx.............................................................................x.xx........xx.xx....................................t.....t...................................t...........................x..........................................................................................................................................................................
Link check done
---------------
Links OK: 1078
Links broken: 43
---------------
Code 404 for http://localhost:3000/veiledninger/kom_i_gang_med_scratch.html in
 - http://localhost:3000/scratch
 - http://localhost:3000/scratch
Code 404 for http://localhost:3000/kart/kart.html in
 - http://localhost:3000/scratch
 - http://localhost:3000/scratch
Code 404 for http://localhost:3000/data/data.html in
 - http://localhost:3000/scratch
 - http://localhost:3000/scratch
Code ENOTFOUND for http://helloworld.py/ in
 - http://localhost:3000/python
 - http://localhost:3000/python
Code 404 for http://localhost:3000/introduksjon_til_web/introduksjon_til_web.html in
 - http://localhost:3000/web
 - http://localhost:3000/web
Code 404 for http://localhost:3000/installasjon/installasjon.html in
 - http://localhost:3000/computercraft
 - http://localhost:3000/computercraft
Code 404 for http://localhost:3000/README.html in
 - http://localhost:3000/learntomod
 - http://localhost:3000/learntomod
Code 404 for http://localhost:3000/scratch/norgestur/norgeskart.png in
 - http://localhost:3000/scratch/norgestur/norgestur
Code 404 for http://localhost:3000/scratch/hvor_i_all_verden/europakart.png in
 - http://localhost:3000/scratch/hvor_i_all_verden/hvor_i_all_verden_1
 - http://localhost:3000/scratch/hvor_i_all_verden/hvor_i_all_verden_2
Code 404 for http://localhost:3000/scratch/hvor_i_all_verden/europakart.zip in
 - http://localhost:3000/scratch/hvor_i_all_verden/hvor_i_all_verden_2
Code 404 for http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_plattform.html in
 - http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_klassikere
Code 404 for http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_klassikere.html in
 - http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_plattform
Code 404 for http://localhost:3000/scratch/fyrverkeri/fyrverkeri_lydogbilder.zip in
 - http://localhost:3000/scratch/fyrverkeri/fyrverkeri
Code 404 for http://localhost:3000/scratch/tegneprogram/tegneprogram_bildefiler.zip in
 - http://localhost:3000/scratch/tegneprogram/tegneprogram
Code 404 for http://localhost:3000/scratch/halloweenimasjon/projiser.html in
 - http://localhost:3000/scratch/halloweenimasjon/halloweenimasjon
Code 404 for http://localhost:3000/scratch/pingviner_pa_tur/bergen.png in
 - http://localhost:3000/scratch/pingviner_pa_tur/pingviner_pa_tur
Code 404 for http://localhost:3000/python/repetisjon/repetisjon.html in
 - http://localhost:3000/python/lister_og_indekser/lister_og_indekser
Code 404 for http://localhost:3000/python/hangman2/hangman.py in
 - http://localhost:3000/python/hangman2/hangman2
 - http://localhost:3000/python/hangman2/hangman2
Code 404 for http://localhost:3000/web/forsvunnet_katt/ressurser/index in
 - http://localhost:3000/web/style_nettsider/style_nettsider
Code 404 for http://localhost:3000/web/animasjon/minecraft_animasjon.zip in
 - http://localhost:3000/web/animasjon/animasjon
Code 404 for http://localhost:3000/web/skjul_ninjaene/ninja.zip in
 - http://localhost:3000/web/skjul_ninjaene/skjul_ninjaene
Code 404 for http://localhost:3000/web/layout/ressurser/layout.zip in
 - http://localhost:3000/web/layout/layout
Code 404 for http://localhost:3000/web/introduksjon_til_web/internett.zip in
 - http://localhost:3000/web/introduksjon_til_web/introduksjon_til_web
 - http://localhost:3000/web/introduksjon_til_web/introduksjon_til_web
 - http://localhost:3000/web/en_hjemmeside/en_hjemmeside
Code 404 for http://localhost:3000/web/forsvunnet_katt/ressurser/catswithhats.gif in
 - http://localhost:3000/web/forsvunnet_katt/forsvunnet_katt
Code 404 for http://localhost:3000/web/index in
 - http://localhost:3000/elm/03_moduler_elm_reactor/03_moduler_elm_reactor
Code 404 for http://localhost:3000/computercraft/hendelser/hendelser.html in
 - http://localhost:3000/computercraft/fjernstyr_en_robot/fjernstyr_en_robot
Code 404 for http://localhost:3000/computercraft/bygg_et_hus/bygg_et_hus.html in
 - http://localhost:3000/computercraft/fjernstyr_en_robot/fjernstyr_en_robot
Code 404 for http://localhost:3000/scratch/veiledninger/kom_i_gang_med_scratch.html in
 - http://localhost:3000/scratch/
 - http://localhost:3000/scratch/
Code 404 for http://localhost:3000/scratch/kart/kart.html in
 - http://localhost:3000/scratch/
 - http://localhost:3000/scratch/
Code 404 for http://localhost:3000/scratch/data/data.html in
 - http://localhost:3000/scratch/
 - http://localhost:3000/scratch/
Code 404 for http://localhost:3000/appinventor/KlaskEnMuldvarp/gress.png in
 - http://localhost:3000/appinventor/KlaskEnMuldvarp/klaskenmuldvarp
Code 404 for http://localhost:3000/appinventor/KlaskEnMuldvarp/mole.png in
 - http://localhost:3000/appinventor/KlaskEnMuldvarp/klaskenmuldvarp
Code 404 for http://localhost:3000/uten_datamaskin/robotvenner/robotvenner_symbolark.pdf in
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
Code 404 for http://localhost:3000/uten_datamaskin/robotvenner/robotvenner_figurer.pdf in
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
Code 404 for http://localhost:3000/uten_datamaskin/robotvenner/robotvenner_trapeser.pdf in
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
Code 404 for http://localhost:3000/scratch/kart/leaflet-image.js in
 - http://localhost:3000/scratch/kart/kart
Code 404 for http://localhost:3000/scratch/kart/kart.js in
 - http://localhost:3000/scratch/kart/kart
Code 404 for http://localhost:3000/scratch/data/papaparse.min.js in
 - http://localhost:3000/scratch/data/data
Code 404 for http://localhost:3000/scratch/data/data.js in
 - http://localhost:3000/scratch/data/data
Code ENOTFOUND for http://myturtle.py/ in
 - http://localhost:3000/python/skilpadder/skilpadder
Code ENOTFOUND for http://xox.py/ in
 - http://localhost:3000/python/tre_pa_rad/tre_pa_rad
Code ENOTFOUND for http://skilpaddetekst.py/ in
 - http://localhost:3000/python/skilpaddetekst/skilpaddetekst
Code 404 for http://ditt-brukernavn.github.io/ in
 - http://localhost:3000/web/publiser/publiser
```

### CODE REVIEW

**Comment by teodorlu** [comment]
I checked and categorized the errors.


# Conclusion

- There are _many_ reported link errors that actually work. If the link checker
  should be of any use, we should make sure we're not flooded with false
  positives. Otherwise, we just learn to ignore it.
- Locally running `./gulp` demands having the `.html` extension, whereas the
  live site doesn't. Any way to have `./gulp` accept the same links as the live
  site?
- I found a few invalid links that I'll fix!

# Error sources

Why the errors appear. List of errors for each type below.

## Type 1: False positive, links from `/category/index.html` are missing category prefix on link targets

Categories of tasks (such as `/scratch/`, `/python/` and `/learntomod/`) are
handled differently by the link checker and the live site. The link checker is
missing the category prefix, and link checking fails. The live site works as
expected.

Link checker links differently from live site.

- Link checker link: `http://localhost:3000/veiledninger/kom_i_gang_med_scratch.html`
- Live site link: `http://localhost:3000/scratch/veiledninger/kom_i_gang_med_scratch.html`

Full list of links (many) below.

## Type 2: False positive?, link checker finds phantom link

Link checker claims `http://localhost:3000/python` links to
`http://helloworld.py/`.

I find this part of the rendered HTML:

```html
<li>Gi programmet ditt navnet helloworld.py og lagre det i den mappen du ønsker.</li>
```

## Type 3: False positive, link seems to work

Link checker fails on `http://localhost:3000/scratch/norgestur/norgeskart.png`.
This works for me.

Some of the files register as downloads. 

## Type 4: Invalid link

These can just be fixed in the their respective tasks; link checker success!

# General comments

Running `./gulp` locally demands linking to tasks with `.html` extension. This
is not required by the live site.

# List of errors

## Type 1

```
Code 404 for http://localhost:3000/veiledninger/kom_i_gang_med_scratch.html in
 - http://localhost:3000/scratch
 - http://localhost:3000/scratch
Code 404 for http://localhost:3000/kart/kart.html in
 - http://localhost:3000/scratch
 - http://localhost:3000/scratch
Code 404 for http://localhost:3000/data/data.html in
 - http://localhost:3000/scratch
 - http://localhost:3000/scratch
...
Code 404 for http://localhost:3000/introduksjon_til_web/introduksjon_til_web.html in
 - http://localhost:3000/web
 - http://localhost:3000/web
Code 404 for http://localhost:3000/installasjon/installasjon.html in
 - http://localhost:3000/computercraft
 - http://localhost:3000/computercraft
Code 404 for http://localhost:3000/README.html in
 - http://localhost:3000/learntomod
 - http://localhost:3000/learntomod

```

## Type 2

```
Code ENOTFOUND for http://helloworld.py/ in
 - http://localhost:3000/python
 - http://localhost:3000/python
...
Code ENOTFOUND for http://myturtle.py/ in
 - http://localhost:3000/python/skilpadder/skilpadder
...
Code ENOTFOUND for http://xox.py/ in
 - http://localhost:3000/python/tre_pa_rad/tre_pa_rad
Code ENOTFOUND for http://skilpaddetekst.py/ in
 - http://localhost:3000/python/skilpaddetekst/skilpaddetekst

```

## Type 3

```
Code 404 for http://localhost:3000/scratch/norgestur/norgeskart.png in
 - http://localhost:3000/scratch/norgestur/norgestur
Code 404 for http://localhost:3000/scratch/hvor_i_all_verden/europakart.png in
 - http://localhost:3000/scratch/hvor_i_all_verden/hvor_i_all_verden_1
 - http://localhost:3000/scratch/hvor_i_all_verden/hvor_i_all_verden_2
Code 404 for http://localhost:3000/scratch/hvor_i_all_verden/europakart.zip in
 - http://localhost:3000/scratch/hvor_i_all_verden/hvor_i_all_verden_2
Code 404 for http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_plattform.html in
 - http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_klassikere
Code 404 for http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_klassikere.html in
 - http://localhost:3000/scratch/veiledninger/kurs_avansert_scratch_plattform
Code 404 for http://localhost:3000/scratch/fyrverkeri/fyrverkeri_lydogbilder.zip in
 - http://localhost:3000/scratch/fyrverkeri/fyrverkeri
Code 404 for http://localhost:3000/scratch/tegneprogram/tegneprogram_bildefiler.zip in
 - http://localhost:3000/scratch/tegneprogram/tegneprogram
Code 404 for http://localhost:3000/scratch/halloweenimasjon/projiser.html in
 - http://localhost:3000/scratch/halloweenimasjon/halloweenimasjon
Code 404 for http://localhost:3000/scratch/pingviner_pa_tur/bergen.png in
 - http://localhost:3000/scratch/pingviner_pa_tur/pingviner_pa_tur
Code 404 for http://localhost:3000/python/repetisjon/repetisjon.html in
 - http://localhost:3000/python/lister_og_indekser/lister_og_indekser
Code 404 for http://localhost:3000/python/hangman2/hangman.py in
 - http://localhost:3000/python/hangman2/hangman2
 - http://localhost:3000/python/hangman2/hangman2
Code 404 for http://localhost:3000/web/animasjon/minecraft_animasjon.zip in
 - http://localhost:3000/web/animasjon/animasjon
Code 404 for http://localhost:3000/web/skjul_ninjaene/ninja.zip in
 - http://localhost:3000/web/skjul_ninjaene/skjul_ninjaene
Code 404 for http://localhost:3000/web/layout/ressurser/layout.zip in
 - http://localhost:3000/web/layout/layout
Code 404 for http://localhost:3000/web/introduksjon_til_web/internett.zip in
 - http://localhost:3000/web/introduksjon_til_web/introduksjon_til_web
 - http://localhost:3000/web/introduksjon_til_web/introduksjon_til_web
 - http://localhost:3000/web/en_hjemmeside/en_hjemmeside
Code 404 for http://localhost:3000/web/forsvunnet_katt/ressurser/catswithhats.gif in
 - http://localhost:3000/web/forsvunnet_katt/forsvunnet_katt
Code 404 for http://localhost:3000/computercraft/hendelser/hendelser.html in
 - http://localhost:3000/computercraft/fjernstyr_en_robot/fjernstyr_en_robot
Code 404 for http://localhost:3000/computercraft/bygg_et_hus/bygg_et_hus.html in
 - http://localhost:3000/computercraft/fjernstyr_en_robot/fjernstyr_en_robot
Code 404 for http://localhost:3000/scratch/veiledninger/kom_i_gang_med_scratch.html in
 - http://localhost:3000/scratch/
 - http://localhost:3000/scratch/
Code 404 for http://localhost:3000/scratch/kart/kart.html in
 - http://localhost:3000/scratch/
 - http://localhost:3000/scratch/
Code 404 for http://localhost:3000/scratch/data/data.html in
 - http://localhost:3000/scratch/
 - http://localhost:3000/scratch/
Code 404 for http://localhost:3000/appinventor/KlaskEnMuldvarp/gress.png in
 - http://localhost:3000/appinventor/KlaskEnMuldvarp/klaskenmuldvarp
Code 404 for http://localhost:3000/appinventor/KlaskEnMuldvarp/mole.png in
 - http://localhost:3000/appinventor/KlaskEnMuldvarp/klaskenmuldvarp
Code 404 for http://localhost:3000/uten_datamaskin/robotvenner/robotvenner_symbolark.pdf in
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
Code 404 for http://localhost:3000/uten_datamaskin/robotvenner/robotvenner_figurer.pdf in
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
Code 404 for http://localhost:3000/uten_datamaskin/robotvenner/robotvenner_trapeser.pdf in
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
 - http://localhost:3000/uten_datamaskin/robotvenner/robotvenner
Code 404 for http://localhost:3000/scratch/kart/leaflet-image.js in
 - http://localhost:3000/scratch/kart/kart
Code 404 for http://localhost:3000/scratch/kart/kart.js in
 - http://localhost:3000/scratch/kart/kart
Code 404 for http://localhost:3000/scratch/data/papaparse.min.js in
 - http://localhost:3000/scratch/data/data
Code 404 for http://localhost:3000/scratch/data/data.js in
 - http://localhost:3000/scratch/data/data

```

## Type 4

```
Code 404 for http://localhost:3000/web/forsvunnet_katt/ressurser/index in
 - http://localhost:3000/web/style_nettsider/style_nettsider
Code 404 for http://localhost:3000/web/index in
 - http://localhost:3000/elm/03_moduler_elm_reactor/03_moduler_elm_reactor
...
Code 404 for http://ditt-brukernavn.github.io/ in
 - http://localhost:3000/web/publiser/publiser

```

**Comment by teodorlu** [comment]
Update: couldn't find anything wrong on the latest revision of the site. Seems like it's all false positives.

File: test-links.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+const http = require('http');
+const package = require('./package.json');
+const Spider = require('node-spider');
+const Static = require('node-static');
+
+/**
+ * Get number of keys in object.
+ * @param {object} obj
+ */
+function length (obj) {
```
Comment (NorwegianKiwi) [inline test-links.js]: Her er det vel bare å gjøre `Object.keys(obj).length`?

File: package.json [hunk 2]
```diff
@@ -14,6 +14,7 @@
     "serve": "ws",
     "servejs": "node ./server.js",
     "test": "mocha --compilers js:babel-core/register --require mock-local-storage --require ignore-styles --require ./test/test_helper.js \"test/**/*@(.js|.jsx)\"",
+    "test:links": "node test-links.js",
```
Comment (NorwegianKiwi) [inline package.json]: Denne burde kanskje legges inn i travis.yml filen? Og muligens i `testall` også?

**Comment by NorwegianKiwi** [comment]
Keep issue #151 open, still need to add this test to travis and testall, and fix links.

File: test-links.js (hunk 1 above)
Comment (arve0) [inline test-links.js]: Ja. Helst på linje 215, kun der den er i bruk.

**Comment by arve0** [comment]
> If the link checker
should be of any use, we should make sure we're not flooded with false
positives. Otherwise, we just learn to ignore it.

Agree. Link test should not give false positives. I've several of the links in https://skagevang.github.io/beta, none are not false positives. @teodorlu Are you mixing up oppgaver.kidsakoder.no (codeclub_lesson_builder), with the new site (codeclub-viewer currently at https://skagevang.github.io/beta/)?

> Locally running ./gulp demands having the .html extension, whereas the live site doesn't. Any way to have ./gulp accept the same links as the live site?

I'm not following you. The link tester is not running `./gulp`. Do you mean `yarn run build`?

By the way, @NorwegianKiwi, is there differences between `yarn run build` and `yarn run build:prod`?

> Type 1: False positive, links from /category/index.html are missing category prefix on link targets

The links are also 404 in the live site:

1. Go to http://oppgaver.kidsakoder.no/beta/scratch
2. Press *Informasjon om kurset*
3. *Kom i gang med Scratch* is 404

This is probably because `http://oppgaver.kidsakoder.no/beta/scratch` is missing `/` in the end, and link in `src/scratch/index.md` is relative. Can be solved in `codeclub-viewer` by adding `/` to the end of url, as is HTTP convention. `folder/index.html` -> `folder/`. And `folder/file.html` -> `folder/file`.

> Type 2: False positive?, link checker finds phantom link

HTML output is `<li>Gi programmet ditt navnet <a href="http://helloworld.py">helloworld.py</a> og lagre det i den mappen du ønsker.</li>`. Avoid this by disallowing `.py` as top level domain: 

```js
md.parser
  .use(attrs)
  .use(headerSections)
  .use(implicitFigures)
  .use(anchor)
  .use(checkbox)
  .linkify.tlds('.py', false)  // linkify: turn of .py top level domain
```
https://github.com/arve0/codeclub_lesson_builder/blob/master/markdown.js#L38

> Type 3: False positive, link seems to work

Similar as type 1, but codeclub-viewer adds another `/` to the relative link `http://oppgaver.kidsakoder.no/beta/scratch/norgestur/--> norgestur <--/norgeskart.png`

> Type 4: Invalid link

These are links in web course, where `.html` is removed by codeclub-viewer.

**Pull request merged**